src/main/host/descriptor/socket/inet/tcp.rs

```diff
@@ -406,6 +409,287 @@ impl LegacyTcpSocket {
+
+                if usize::try_from(optlen).unwrap() < std::mem::size_of::<OptType>() {
+                    return Err(Errno::EINVAL.into());
+                }
+
+                let optval_ptr = TypedPluginPtr::new::<OptType>(optval_ptr, 1);
+                let val: u64 = memory_manager.read_vals::<_, 1>(optval_ptr)?[0]
+                    .try_into()
+                    .or(Err(Errno::EINVAL))?;
+
+                // linux kernel doubles this value upon setting
+                let val = val * 2;
+
+                // Linux also has limits SOCK_MIN_SNDBUF (slightly greater than 4096) and the sysctl
+                // max limit. We choose a reasonable lower limit for Shadow. The minimum limit in
+                // man 7 socket is incorrect.
+                let val = std::cmp::max(val, 4096);
+
+                // This upper limit was added as an arbitrarily high number so that we don't change
+                // Shadow's behaviour, but also prevents an application from setting this to
+                // something unnecessarily large like INT_MAX.
+                let val = std::cmp::min(val, 268435456); // 2^28 = 256 MiB
+
+                unsafe { c::legacysocket_setOutputBufferSize(self.as_legacy_socket(), val) };
+                unsafe { c::tcp_disableSendBufferAutotuning(self.as_legacy_tcp()) };
+            }
+            (libc::SOL_SOCKET, libc::SO_RCVBUF) => {
+                type OptType = libc::c_int;
+
+                if usize::try_from(optlen).unwrap() < std::mem::size_of::<OptType>() {
+                    return Err(Errno::EINVAL.into());
+                }
+
+                let optval_ptr = TypedPluginPtr::new::<OptType>(optval_ptr, 1);
+                let val: u64 = memory_manager.read_vals::<_, 1>(optval_ptr)?[0]
+                    .try_into()
+                    .or(Err(Errno::EINVAL))?;
+
+                // linux kernel doubles this value upon setting
+                let val = val * 2;
+
+                // Linux also has limits SOCK_MIN_RCVBUF (slightly greater than 2048) and the sysctl
+                // max limit. We choose a reasonable lower limit for Shadow. The minimum limit in
+                // man 7 socket is incorrect.
+                let val = std::cmp::max(val, 2048);
+
+                // This upper limit was added as an arbitrarily high number so that we don't change
+                // Shadow's behaviour, but also prevents an application from setting this to
+                // something unnecessarily large like INT_MAX.
+                let val = std::cmp::min(val, 268435456); // 2^28 = 256 MiB
+
+                unsafe { c::legacysocket_setInputBufferSize(self.as_legacy_socket(), val) };
+                unsafe { c::tcp_disableReceiveBufferAutotuning(self.as_legacy_tcp()) };
+            }
+            (libc::SOL_SOCKET, libc::SO_REUSEADDR) => {
+                // TODO: implement this, tor and tgen use it
+                log::trace!("setsockopt SO_REUSEADDR not yet implemented");
```

Comment:
Should these unsupported cases return errors? Or does that break our current simulations using tgen and tor?

Reply:
It probably breaks tor or tgen, but I'm not sure. It has always been this way (from 2020):

https://github.com/shadow/shadow/blob/865c384824280c08cc8f8a76afef1285a72970c5/src/main/host/syscall/socket.c#L394-L405